### PR TITLE
Swapping Context and Constr and defining declarations on constr in Constr

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** architecture

The motivation for this PR is to eventually be able to use contexts of declarations in the definition of the `Case` constructor of `constr` (as discussed in the last steps of CEP [#34](https://github.com/coq/ceps/pull/34)).

It can also be seen as moving `Context.Rel.Declaration.t` and the 5 other `Context.*.t` of `context.ml` (i.e. the instantiation of the generic `pt` to `constr`/`types`) to a place where they are less isolated that they became after the generalization of `Context` to support `EConstr`. With this move, `Context` becomes exclusively polymorphic.

From the API point of view, this implies that `Constr` now includes `Context` and that the `t` types of `Context` are now only available in `Constr.Context.Rel.Declaration` and co and not any
more in `Context.Rel.Declaration`. So, more `open Constr` are explicitly needed (I hope this does not go against other ongoing design choices).

The inclusion of `Context` in `Constr` is done to the price of a pretty heavy boilerplate. Typically, I would just have wanted to be able to say:
```
include Context
  extending Rel.Declaration with t = (constr, types) Rel.Declaration.pt
  extending Rel with t = Declaration.t list
  extending Named.Declaration with t = (constr, types) Named.Declaration.pt
  extending Named with t = Named.Declaration.t list
  extending Compacted.Declaration with t = (constr, types) Compacted.Declaration.pt
  extending Compacted with t = Compacted.Declaration.t list
```
but I did not find how to do it concisely.

Note: I did not put a kind label. Maybe `redesign` is the closest one that I could found but it is not about a feature, it is purely architectural.

### CODE REVIEW

**Review by ppedrot — APPROVED** [review]
I like the idea. The implementation is a bit heavyweight indeed, but that's a one-time shot, and the compiler should help maintenance anyways.

**Review by ejgallego — DISMISSED** [review]
Please go through a proper deprecation phase.

**Comment by ejgallego** [comment]
On an additional thought, it seems to me that we have been doing modifications to core APIs [the renaming of context, the split of `Term`, etc...] a bit lightly, but those changes have created _huge_ pain to 3rd party plugin developers.

So I am not sure I'd like to support more modification to the core organization without going first thru a CEP phase to try to understand what the global picture is.

**Comment by SkySkimmer** [comment]
Instead of doing this inclusion dance why not just have in constr.ml `type rel_context = (constr, constr) Context.Rel.pt`?

**Comment by Zimmi48** [comment]
@ejgallego It's funny that you are the one being concerned about it after having been the one to propose the biggest API changes in the past releases... (but I know you have also helped a lot of plugin writers port their plugins, so you know what their life must be.) So indeed, it is important to consider what we inflict on plugin writers. On the other hand, we can't suddenly decide that there will be no more API changes without a deprecation period. I'd support this objective medium-term but this has to be considered carefully.

File: dev/doc/changes.md
Comment (Zimmi48) [inline dev/doc/changes.md]: This is a markdown file so if you don't put backquotes around this, the star is interpreted as marking the begin of an emphasis.

**Comment by ejgallego** [comment]
> It's funny that you are the one being concerned about it after having been the one to propose the biggest API changes in the past releases... (but I know you have also helped a lot of plugin writers port their plugins, so you know what their life must be.)

Well, indeed I am saying that because I happen to know what I am talking about in this case. In fact, my changes were discussed over several hundredth messages and several WG [including talks to present them/demos], and a CEP of course https://github.com/coq/ceps/pull/19.

They caused low-to-zero pain to plugin writers, and provided some other advantages.

However, the changes I am talking about here don't meet any of these conditions, and worse, they have introduced consistency issues in the past. So I'd like to have a global picture to understand what the direction is.

> So indeed, it is important to consider what we inflict on plugin writers. On the other hand, we can't suddenly decide that there will be no more API changes without a deprecation period. I'd support this objective medium-term but this has to be considered carefully.

Note that my review and commentary are in different comments, that mean that they don't have the same level of strength. That being said, the current "unofficial" policy is that "no deprecation" only applies that parts of the API that indeed doesn't admit a deprecation phase [for example it'd be crazy to do the AST stuff with a deprecation layer]. This PR is not in that case, as there is no reason to remove the aliases, so my X stands for now.

**Comment by herbelin** [comment]
> Instead of doing this inclusion dance why not just have in constr.ml type rel_context = (constr, constr) Context.Rel.pt?

I'd be fine with that. I used a `t` type to be compliant with some relatively spread usage these days to prefer names of the form `MyConcept.t` over `my_concept`.

> but those changes [to the API] have created huge pain to 3rd party plugin developers.

I indeed know that changes to the API induce pain to 3rd party plugin developers (and of course not only in the case of Coq). There is a middle-term to find which I hope shall be found.

In the case of the API change here, it is relatively minor (I would say more minor for instance than the split of `Command`), but indeed, the question of whether the API should be stabilized and all new changes be compatible has to be raised, probably also involving 3rd party plugin developers.

About a deprecation, I would be ok to do so but what forms would it take? I cannot have both `Context` come after `Constr` and before `Constr` so, these `t` have to be in one of those only, or do I miss something?

File: dev/doc/changes.md
Comment (herbelin) [reply dev/doc/changes.md]: Thanks for noticing. Fixed.

**Comment by ejgallego** [comment]
> About a deprecation, I would be ok to do so but what forms would it take? I cannot have both Context come after Constr and before Constr so, these t have to be in one of those only, or do I miss something?

Can't you just keep the old types as deprecated aliases? I'll have a look ASAP.

**Comment by Zimmi48** [comment]
Note also that the fix to plugins is (apparently) just a simple `open Constr`...

**Comment by ejgallego** [comment]
> Note also that the fix to plugins is (apparently) just a simple open Constr...

The advantage of keeping the deprecated aliases is that plugin developers get a warning telling them what to do instead. This is almost infinitely more useful than the experience of "my code doesn't compile, this is not there, where did it move?"

For one reason or other dev documentation is still not very discoverable IMHO. This is of course related to your #6137 among others.

**Comment by ejgallego** [comment]
Who's the one to drive this PR?

**Comment by ejgallego** [comment]
> As for a deprecation phase, I don't know how to do one. It is a situation similar to the split of command.ml into several files: since command.ml disappeared, there is no place to put a deprecation warning.

I would be OK driving it, but this seems like kernel stuff so I can only do with the blessing of kernel maintainers; `command.ml` was mostly a private API, no plugin was impacted, this however may impact every plugin so more care is needed IMVHO.

I'll have a look to see if the deprecation stub is possible to write, I think it should be possible.

**Comment by maximedenes** [comment]
> I can only do with the blessing of kernel maintainers

I'll try to have a look quickly, but I want to clear 8.8.0 release blockers from my list (and that includes the remaining bits of the Sphinx migration).

**Comment by SkySkimmer** [comment]
We could do a deprecation phase by having the definition of context in constr.ml
```ocaml
(* constr.ml *)
type constr 

module Context = ...

(*****************************)
(* context.ml *)
include Constr.Context

(*****************************)
(* context.mli *)
(* same as now but with monomorphised types deprecated *)
```

**Comment by ejgallego** [comment]
Note that indeed @herbelin is right about `Command`, I could have kept `comman.ml` re-exporting the old functions indeed as for compatibility. I guess I didn't do as I really thought this module was not meant to be used by plugins, but I can prepare a PR for 8.8 if you think we should ship it.

**Comment by SkySkimmer** [comment]
>I'm still open to the type rel_context = (constr, constr) Context.Rel.pt suggestion if we accept that the Foo.t model is not to be exclusive.

I think rel_context is better than the PR-current include magic.

**Comment by herbelin** [comment]
What to do? Everyone seems pretty busy, and, at the same time, more opinions would have to be expressed so that one can have an idea of where the balance tips.

@ejgallego, @maximedenes: would you have time to have a look as you wanted to do or are you really overbusy?

@skyskimmer:
> We could do a deprecation phase by having the definition of context in constr.ml

I don't understand your proposal. The incompatibility is that `Context.Rel.t` and cie are now `Constr.Context.Rel.t`and cie  (or `Constr.rel_context` with your proposal). I don't see a way to have
`Context.Rel.t` available with the swap.

**Comment by SkySkimmer** [comment]
Like this dad0becc2d.
It's not very pretty and I'm dubious of it being worth doing but it's possible.

**Comment by ejgallego** [comment]
I have dismissed the review as if in the end we cannot do a deprecation, that's life.

On the other hand I am not knowledgeable enough as to judge CEP#34

**Comment by herbelin** [comment]
@skyskimmer: It makes `constr.ml` big but why not.

@ejgallego: the ultimate idea with coq/ceps#34 is that the branches and return predicate of a `Cases` are made of a `rel_context` and a term. But even w/o going this far, providing combinators working on branches and return predicate with full binders needs to know `Context.Rel.Declaration.pt` in `constr.ml` (or to move these combinators outside of `constr.ml`, but we then lose centralization of the API).

Another solution would be to define only the `pt` types in `Constr` and to keep the rest of the API in `Context`.

Or to have a `cContext.ml` file coming before `constr.ml` and do @skyskimmer trick to preserve the original `Context` namespace.

But both latter solutions would need the complex module extension trick. Then, this could be combined with using `rel_context` and `rel_declaration` instead of `Context.Rel.t` and `Context.Rel.Declaration.t`. (But then, we lose again the compatibility with former `Context.Rel.t`.) 

How to order priorities here? Maybe adopting @skyskimmer's proposal, even if it makes `constr.ml` bigger?

**Comment by ejgallego** [comment]
IMVHO if the problem for "bigger" is "kernel size", I think we are past the era that is so important as to triumph over other issues. Thanks with the explanation, my head is a bit overloaded these days so I am not sure I can think cleanly, I guess I'd defer feedback to other developers until I recover a bit of mental clarity.

**Comment by herbelin** [comment]
BIgger only means including context.ml in constr.ml. I'm eventually adding a couple of new (canonical) combinators, thus making the kernel bigger (*) but that's not the issue in this PR.

(*) We cannot exclude that there are typos in these combinators, but, morally, they should be mechanically generated from the structure of terms, so there is no algorithmic nor logic-relevant issues with these combinators.

These new combinators (in #7109) are not making the kernel a lot more fragile than say, using an optimized implementation of `List.append` (the main risk I may see is with rarer constructors like say `CoFIx` toegther with a complex structure of binders, where a typo might get undetected by daily code).

Moreover, it seems easier to me to visually inspect the regularity of the code of these combinators than, say, to check that the optimized imperative code of `List.append` has no problems in degenerate cases.

@SkySkimmer:
> My preferred solution is that we don't have a copy of Context in Constr, we just have Constr.rel_context and don't bother with a deprecation phase.

So, you mean as it is currently in the PR but without the `module` extension stuff, and just a `Constr.rel_context` outside of the `Context` name space?

**Comment by SkySkimmer** [comment]
Yes

Gaëtan Gilbert

**Comment by herbelin** [comment]
> > So, you mean as it is currently in the PR but without the module extension stuff, and just a Constr.rel_context outside of the Context name space?

>  Yes

@ppedrot, would that be ok for you?

**Comment by ppedrot** [comment]
@herbelin Fine with me.

**Comment by herbelin** [comment]
I pushed a new version then without the extra `open Constr`, without the intricate module extension, but with 6 type renamings as shown in `changes.md`. @SkySkimmer, @ppedrot, @ejgallego, please confirm that this is what you had in mind.

According to the [assignment rules](https://github.com/coq/coq/blob/master/dev/doc/MERGING.md), @maximedenes should self-assign himself as primary kernel maintainer, right?

**Comment by ejgallego** [comment]
I have no real opinion, that seems fine to me; it is a bit unfortunate that we had plugins writers to go through the pain of doing the module change and now undoing it, but yeah, if we are convinced this is the fnal choice let's apologize and move on.

**Review by SkySkimmer — APPROVED** [review]
Yes, that's what I meant.

**Comment by herbelin** [comment]
@ejgallego: note the new echo of `Constr.named_context` to the existing `EConstr.named_context` which is not unpleasant.

**Comment by herbelin** [comment]
Overlays there. Rebase done.

**Comment by herbelin** [comment]
Taking a concrete decision on this PR would allow to make progresses towards coq/ceps#34. As far as I understand, if I summarize correctly, there was anyway an agreement on the PR, even though it is a change of API. So, the ball is the court of the assignee.

@maximedenes, would you have a little bit of time available in between your multiple engagements?

**Review by maximedenes — DISMISSED** [review]
As it is, this PR brings back `Pre_env`, it shouldn't. Probably due to a conflict, sorry about that.

Apart from that, the PR looks ok, but I'm a bit puzzled by the renaming we are going to ask plugin writers to do:

`Context.Rel.t` into `Constr.rel_context`
`Context.Named.t` into `Constr.named_context`
`Context.Compacted.t` into `Constr.compacted_context`
`Context.Rel.Declaration.t` into `Constr.rel_declaration`
`Context.Named.Declaration.t` into `Constr.named_declaration`
`Context.Compacted.Declaration.t` into `Constr.compacted_declaration`

I thought we were going in the other direction (`name` into `Name.t`) these days. In fact, in the precise case of contexts, didn't we do such renamings in the other way around recently?

**Comment by mattam82** [comment]
AFAIU, we agreed for the short aliases and renaming on the user side. IMHO Context.Rel.Declaration.t and co was slightly overkill qualification for such a widely used structure. I’d be curious to know how much pain that is for @SkySkimmer whose sprop branch must modify contexts.

**Comment by SkySkimmer** [comment]
A change in these types' names should not impact sprop much if at all.

**Comment by herbelin** [comment]
> As it is, this PR brings back Pre_env, it shouldn't. Probably due to a conflict, sorry about that.

Indeed. I failed to see that it was re-added. Now fixed.

> I thought we were going in the other direction (name into Name.t) these days. In fact, in the precise case of contexts, didn't we do such renamings in the other way around recently?

Are you then satisfied with @mattam82's answer (and by the discussion above)?

**Comment by maximedenes** [comment]
> Indeed. I failed to see that it was re-added. Now fixed.

Thanks!

> Are you then satisfied with @mattam82's answer (and by the discussion above)?

I personally don't care much about these naming conventions, so as long as others are happy, I'm happy. @ejgallego, are you ok with the final shape of the renamings?

**Comment by ejgallego** [comment]
> @ejgallego, are you ok with the final shape of the renamings?

Sure, this is a hard problem and I don't want to obstruct, anyways most opinions here are subjective.

But I do indeed wonder if we should spend a bit more time thinking about the naming policy, the two latest big changes `Context` and `Names` went in undiscussed, so it is good we are trying to discuss more.

In particular, one motivation for having the `Name.t` style is that it allows for easy use of `Name` as a module parameter provided it implementes `Name.equal` etc... So why don't we have indeed `Context.t`

Maybe `Context` should be a fucntor? I dunno; not an easy choice. Certianly having `Context.print`, `Context.equal` etc... would be desirable. Maybe `Context.t` should take a type parameter?

 BTW, there was some discussion about this in https://github.com/coq/ceps/pull/19

**Comment by maximedenes** [comment]
> But I do indeed wonder if we should spend a bit more time thinking about the naming policy, the two latest big changes Context and Names went in undiscussed, so it is good we are trying to discuss more.

Concretely, are you requesting more time? Or should I merge now?

**Comment by ejgallego** [comment]
Umm, maybe I'd like to have another look at this tomorrow if possible.

**Comment by ejgallego** [comment]
Having a look now.

**Comment by ejgallego** [comment]
Ok, so I think that the patch is OK for me, however I would **very strongly** encourage that if we are going to clean contexts up we take a step further and remove the current multi-module setup in favor a couple of modules "Declaration" and "Context" that are properly polymorphic. It just makes little sense to me to have the current number of types instead of a single polymorphic one.

As far as I can see the only bits we'd care about are the smart mappers but that shouldn't be a problem. Of course I didn't try to do it so I am not sure my proposal is feasible, but I don't see why it should not work.

WDYT folks?

**Comment by herbelin** [comment]
> It just makes little sense to me to have the current number of types instead of a single polymorphic one.

You mean a common type for `rel`, `named`, `compacted`? No objection on my side (modulo the sempiternal question of the compromise between cleaning and stability).

**Comment by ejgallego** [comment]
> You mean a common type for rel, named, compacted? No objection on my side (modulo the sempiternal question of the compromise between cleaning and stability).

What do others think, @ppedrot ? You are sometimes opposed to the use of polymorphic types?

**Comment by ejgallego** [comment]
> You mean a common type for rel, named, compacted? No objection on my side (modulo the sempiternal question of the compromise between cleaning and stability).

Indeed, that's what I mean, adding an additional type parameter. So `map` etc become now `Context.map` instead of `Context.Rel.` etc...

We could provide a pretty good compatiblity layer here in the way I did for misctypes, I think breakage shouldn't be larger than what already this PR does, but of course I am just writing code in my head so I could be wrong.

**Comment by ppedrot** [comment]
> What do others think, @ppedrot ? You are sometimes opposed to the use of polymorphic types?

This really depends on the kind of code involved. Here it is very natural to have contexts polymorphic over terms.

**Comment by ejgallego** [comment]
> Here it is very natural to have contexts polymorphic over terms.

I was refeering to the index parameter actually, but indeed I think it is natural too; but it is not trivial change so we should be sure we all think it makes sense IMO.

**Comment by mattam82** [comment]
So we should have a context structure polymorphic over names and terms? I’d
be ok with that. I’m not sure what “index parameter” refers to @ejgallego
IIUC we would have one structure for all local contexts ne they named, rel,
compacted, constr or econstr based.

**Comment by ejgallego** [comment]
> IIUC we would have one structure for all local contexts ne they named, rel,
> compacted, constr or econstr based.

Yeah that'd be the idea; from my quick look most operations would be made polymoprhic; the only tricky part is what to do about equality; there are not so many functions using it but indeed if name is polymorhpic then the `Name.equal` / `Id.equal` have to be determined somehow; and it is not clear to me if a functor or a param would be better.

**Comment by mattam82** [comment]
I guess a couple params would do here.

**Comment by herbelin** [comment]
Any plans to experiment a polymorphic type of contexts in the short term? Or any consensus that such a change is worth?

From my point of view, the question is that I started to develop work on top of this PR to address coq/ceps#34, making the `Constr.Case` node depending on `rel_context`, so as to fix the confusion between a `match` branch and a lambda and between a `return` predicate and a lambda. That would be a pity that this other work is blocked by an undecidable question arising on the way, without timeout to turn it into a decided one.

If ever the idea of changing the types of context to a single polymorphic type makes clearly its way, I can work on it a posteriori say in September.

Then, would that be ok for everyone to proceed with this PR, living open the possibility of further changes in `Context` for later?

**Comment by ppedrot** [comment]
@herbelin I'm fine with this PR for now, we could defer other experiments later on.

**Comment by ejgallego** [comment]
@herbelin I am OK if you want this to be merged as-is.

**Comment by mattam82** [comment]
Ok as well.

**Comment by herbelin** [comment]
@maximedenes: it seems to be in your hands.

**Review by maximedenes — APPROVED** [review]

**Comment by maximedenes** [comment]
@herbelin The last thing that is needed is a rebase of Equations overlay (there is no conflict, we simply need to include other changes done since then). @gares already took care of rebasing the elpi overlay. I'll rerun the corresponding jobs and merge.

**Comment by herbelin** [comment]
Equations overlay rebased.

**Pull request merged**
